rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
-    let len_bytes = u32::to_be_bytes(len as u32);
-    let out: &mut [u8; 3] = (&mut bytes[len_offset..len_offset + 3])
-        .try_into()
-        .unwrap();
-    out.copy_from_slice(&len_bytes[1..]);
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
 }
 
-pub fn read_vec_u8<T: Codec>(r: &mut Reader) -> Option<Vec<T>> {
-    let mut ret: Vec<T> = Vec::new();
-    let len = usize::from(u8::read(r)?);
-    let mut sub = r.sub(len)?;
+impl<L, T> DerefMut for TlsVec<L, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.items
+    }
+}
 
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
+impl<L, T> From<TlsVec<L, T>> for Vec<T> {
+    fn from(tls: TlsVec<L, T>) -> Self {
+        tls.items
+    }
+}
+
+impl<L, T> From<Vec<T>> for TlsVec<L, T> {
+    fn from(items: Vec<T>) -> Self {
+        Self {
+            items,
+            len: PhantomData,
+        }
     }
+}
 
-    Some(ret)
+impl<L, T> Default for TlsVec<L, T> {
+    fn default() -> Self {
+        Self {
+            items: Default::default(),
+            len: Default::default(),
+        }
+    }
 }
 
-pub fn read_vec_u16<T: Codec>(r: &mut Reader) -> Option<Vec<T>> {
-    let mut ret: Vec<T> = Vec::new();
-    let len = usize::from(u16::read(r)?);
-    let mut sub = r.sub(len)?;
+const MAX_VEC_BYTES: usize = 0x1_0000;
 
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
+#[allow(clippy::use_self)]
+impl LengthPrefix for u8 {
+    fn write_into(len: usize, offset: usize, bytes: &mut [u8]) {
+        debug_assert!(len <= 0xff);
+        bytes[offset] = len as u8;
     }
 
-    Some(ret)
+    const BYTES: usize = 1;
+    const ZERO_LEN: &'static [u8] = &[0];
 }
 
-pub fn read_vec_u24_limited<T: Codec>(r: &mut Reader, max_bytes: usize) -> Option<Vec<T>> {
-    let mut ret: Vec<T> = Vec::new();
-    let len = u24::read(r)?.0 as usize;
-    if len > max_bytes {
-        return None;
+#[allow(clippy::use_self)]
+impl LengthPrefix for u16 {
+    fn write_into(len: usize, offset: usize, bytes: &mut [u8]) {
+        debug_assert!(len <= 0xffff);
+        let out: &mut [u8; 2] = (&mut bytes[offset..offset + 2])
+            .try_into()
+            .unwrap();
+        *out = Self::to_be_bytes(len as u16);
     }
 
-    let mut sub = r.sub(len)?;
+    const BYTES: usize = 2;
+    const ZERO_LEN: &'static [u8] = &[0, 0];
+}
 
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
+impl LengthPrefix for u24 {
+    fn write_into(len: usize, offset: usize, bytes: &mut [u8]) {
+        debug_assert!(len <= 0xff_ffff);
+        let len_bytes = u32::to_be_bytes(len as u32);
+        let out: &mut [u8; 3] = (&mut bytes[offset..offset + 3])
+            .try_into()
+            .unwrap();
+        out.copy_from_slice(&len_bytes[1..]);
     }
 
-    Some(ret)
+    const BYTES: usize = 3;
+    const ZERO_LEN: &'static [u8] = &[0, 0, 0];
+}
+
+pub trait LengthPrefix: Codec + Copy + Debug {
```

Comment:
In general, traits should be defined ahead of their implementations.

Let's document the trait.

Given the code above, it is possible to construct a `TlsVec` that will fail to encode because the length of the vector is larger than can be encoded. This will be discovered only during encoding of the `TlsVec` and not before. This in particular should be documented because it is surprising.